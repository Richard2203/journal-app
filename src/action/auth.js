import { types } from '../types/types';
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { startLoading, finishLoading } from './ui';

//? La manera de manejar los dispatch es centralizar en un archivo las acciones
//? que existen

export const startLoginEmailPassword = (email, password) => {
	// el DOM ofrecera el dispatch y entonces lo ejecutara
	return (dispatch) => {
		dispatch(startLoading());

		// signInWithEmailAndPassword(email, password) recibe las credenciales
		// y las compara con las existentes en la BD; retorna una promesa
		firebase
			.auth()
			.signInWithEmailAndPassword(email, password)
			.then(({ user }) => {
				// este dispatch actualiza el Redux
				dispatch(login(user.uid, user.displayName));

				dispatch(finishLoading());
			})
			.catch((e) => {
				console.log(e);
				dispatch(finishLoading());
			});
	};
};

export const startRegisterUser = (email, password, name) => {
	return (dispatch) => {
		// la funcion createUserWithEmailAndPassword(email,password)
		// crea un usuario en firebase con las credenciales proporcionadas
		// por el usuari; y retorna una promesa
		firebase
			.auth()
			.createUserWithEmailAndPassword(email, password)
			.then(async ({ user }) => {
				// usamos el metodo updateProfile y le enviamos un nombre
				// para poder asignar el nombre del usuario a la cuenta
				await user.updateProfile({ displayName: name });

				// este dispatch() registra el usuario en nuestro Redux
				dispatch(login(user.uid, user.displayName));
			})
			.catch((e) => {
				console.log(e);
			});
	};
};

// el login con un proveedor es asincrono por lo cual debemos retornar
// un callback
export const startGoogleLogin = () => {
	return (dispatch) => {
		// retorna una promesa por lo cual lo manejaremos con then
		firebase
			.auth()
			.signInWithPopup(googleAuthProvider)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
			});
	};
};

// para manejar el dispatch se crea un metodo que retorne el tipo de accion
// y los datos
export const login = (uid, name) => ({
	type: types.login,
	payload: {
		uid,
		name,
	},
});

export const startLogout = () => {
	return async (dispatch) => {
		// esperamos a que firebase termine de hacer logout del usuario
		await firebase.auth().signOut();
		dispatch(logout());
	};
};

export const logout = () => ({ type: types.logout });
