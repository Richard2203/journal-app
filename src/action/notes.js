import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpLoad } from '../helpers/fileUpLoad';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

export const startNewNote = () => {
	// ambos argumentos son proveidos por el DOM, el primero es el dispatch
	// y el segundo nos permite obtener el state
	return async (dispatch, getState) => {
		// obtenemos la propiedad ui del objeto auth
		const { uid } = getState().auth;

		const newNote = {
			title: '',
			body: '',
			date: new Date().getTime(),
		};

		// BD FIREBASE - es una BD no relacional, funciona colocando
		// colecciones -> documentos -> colecciones -> documentos -> etc

		// db.collection(<path:string>) retorna una referencia a la collecion
		// de la bd.
		// recibe el path donde guardara la nueva informacion; para agregar un
		// nuevo elemento se emplea el metodo add(<dato>)
		const dbRef = await db.collection(`${uid}/journal/notes`).add(newNote);

		// enviando a dispatche la nueva nota, le pasamos el id y el objeto
		// de la nueva nota
		dispatch(activeNote(dbRef.id, newNote));
	};
};

export const activeNote = (id, note) => ({
	type: types.notesActive,
	payload: { id, ...note },
});

export const startLoadingNotes = (uid) => {
	return async (dispatch) => {
		// -Disparo el metodo loadNotes para cargar las notas del usuario
		//  se hace aqui por ser el primer lugar donde se obtiene el user.
		// -Se emplea await puesto que loadNotes retorna una promesa
		const notes = await loadNotes(uid);

		// enviando al Redux las notas cargadas de firebase
		dispatch(setNotes(notes));
	};
};

export const setNotes = (notes) => ({
	type: types.notesLoad,
	payload: notes,
});

export const startSaveNote = (note) => {
	return async (dispatch, getState) => {
		const { uid } = getState().auth;

		// firebase no permite guardar campos "undefined" por lo cual indicamos
		// si nuestro objeto "note" tiene "undefined" en la propieadad "url"
		// entonces lo elimina
		!note.url && delete note.url;

		// hacemos una copia de "note" para evitar modificar
		// accidentalmente "note" puesto que los objeto se manejan por
		// referencia
		const noteToFirestore = { ...note };

		// con "delete" borramos la propiedad id del objeto noteToFirestore
		delete noteToFirestore.id;

		// .doc() nos devuelve una referencia al documento para poderlo manipular
		// este metodo recibe un path
		// .update() recibe un objeto con la informacion a modificar
		await db
			.doc(`${uid}/journal/notes/${note.id}`)
			.update(noteToFirestore)
			.catch((e) => Swal.fire('Not Saved', e, 1500));

		// enviando la nota actualizada a Redux
		dispatch(refreshNote(note.id, note));
		Swal.fire('Saved', note.title, 'success');
	};
};

export const refreshNote = (id, note) => ({
	type: types.notesUpdated,
	payload: { id, note: { id, ...note } },
});

export const startUpLoading = (file) => {
	return async (dispatch, getState) => {
		const { active: noteActive } = getState().notes;

		// muestra la ventana pero bloquea la opcion de cerrarla
		Swal.fire({
			title: 'Uploading',
			text: 'please wait',
			allowOutsideClick: false,
			showConfirmButton: false,
			willOpen: () => {
				Swal.showLoading();
			},
		});

		const fileUp = await fileUpLoad(file);
		noteActive.url = fileUp;

		dispatch(startSaveNote(noteActive));

		// cierra a ventana Swal
		Swal.close();
	};
};
