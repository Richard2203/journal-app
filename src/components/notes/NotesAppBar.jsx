import React from 'react';
import { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
	startDeleting,
	startSaveNote,
	startUpLoading,
} from '../../action/notes';

const NotesAppBar = () => {
	const dispatch = useDispatch();
	const refFileInput = useRef(null);
	const { active: note } = useSelector((state) => state.notes);

	const handleSave = () => {
		dispatch(startSaveNote(note));
	};

	const handlePictureClick = () => {
		// usando el hook useRef para hacer referencia a input de tipo
		// "file" y controlarlo en esta funcion
		refFileInput.current.click();
	};

	const handleFileChange = (e) => {
		const file = e.target.files[0];
		file && dispatch(startUpLoading(file));
	};

	const handleDeleteNote = () => {
		dispatch(startDeleting(note.id));
	};

	return (
		<div className="notes__appbar">
			<span>28 de noviembre del 2050</span>
			<div>
				<input
					type="file"
					name="file"
					id="fileImage"
					ref={refFileInput}
					onChange={handleFileChange}
					style={{ display: 'none' }}
				/>
				<button className="btn cursor" onClick={handleDeleteNote}>
					Borrar Nota
				</button>
				<button className="btn cursor" onClick={handlePictureClick}>
					Picture
				</button>
				<button className="btn cursor" onClick={handleSave}>
					Save
				</button>
			</div>
		</div>
	);
};

export default NotesAppBar;
