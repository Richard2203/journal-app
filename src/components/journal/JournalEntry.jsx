import React from 'react';
// esta libreria nos permite la manipulacion de fechas
// se instala con el comando
// 	npm install moment --save
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../action/notes';

const JournalEntry = ({ id, date, title, body, url }) => {
	const dateMoment = moment(date);

	const dispatch = useDispatch();

	const handleEntryClick = () => {
		dispatch(activeNote(id, { date, title, body, url }));
	};

	return (
		<div className="journal__entry cursor" onClick={handleEntryClick}>
			{/* en React la etiqueta style se maneja pasando las propieades
            como si de un objeto se tratara */}
			{url && (
				<div
					className="journal__entry-picture"
					style={{
						backgroundSize: 'cover',
						backgroundImage: `url(${url})`,
					}}
				></div>
			)}
			<div className="journal__entry-body">
				<p className="journal__entry-title">{title}</p>
				<p className="journal__entry-content">{body}</p>
			</div>

			<div className="journal__entry-date">
				{/* el formato de las fechas se optiene de la documentacion de 
				https://momentjs.com/ */}
				<span>{dateMoment.format('dddd')}</span>
				<h4>{dateMoment.format('Do')}</h4>
			</div>
		</div>
	);
};

export default JournalEntry;
