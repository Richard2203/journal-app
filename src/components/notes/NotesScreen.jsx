import React, { useEffect } from 'react';
import { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote } from '../../action/notes';
import { useForm } from '../../hooks/useForm';
import NotesAppBar from './NotesAppBar';

const NotesScreen = () => {
	const dispatch = useDispatch();

	const { active: note } = useSelector((state) => state.notes);

	// useForm maneja un unico estado y es por eso que aun cuando le mandamos
	// un nuevo "note" este no modificara su estado original; para modificarlo
	// debemos emplear un metodo similar a handleInputChange que modifique todo
	// el estado
	const [formValues, handleInput, reset] = useForm(note);
	const { body, title } = formValues;

	// useRef permite almacenar una variable mutable que no redibujara al
	// componente si su valor cambia
	// useRef apuntara a note.id, de tal modo que
	const activeId = useRef(note.id);

	// si el Id de la nota activa obtenida del selector es diferente de  activeId
	// del useRef entonces resetea el form por la nueva nota
	useEffect(() => {
		if (note.id !== activeId.current) {
			reset(note);
			// la nota vieja pasa a tener el id de la nota nueva
			activeId.current = note.id;
		}
	}, [note, reset]);

	useEffect(() => {
		dispatch(activeNote(formValues.id, { ...formValues }));
	}, [formValues, dispatch]);

	return (
		<div className="notes__main-content">
			<NotesAppBar />

			<div className="notes__content">
				{note.url && (
					<div className="notes__images">
						<img
							src="https://images.alphacoders.com/111/thumb-1920-1114657.jpg"
							alt="Demon Slayer"
						/>
					</div>
				)}

				<div className="form">
					<div className="form-control">
						<input
							type="text"
							id="title"
							name="title"
							className="form-control__input"
							autoComplete="off"
							placeholder=" "
							value={title}
							onChange={handleInput}
						/>
						<label htmlFor="title" className="form-control__label">
							Some Awesome title
						</label>
					</div>

					<div className="form-control">
						<textarea
							id="body"
							name="body"
							className="form-control__area"
							style={{ height: '10rem' }}
							placeholder=" "
							value={body}
							onChange={handleInput}
						></textarea>

						<label
							htmlFor="body"
							className="form-control__label-area"
						>
							what happened today?
						</label>
					</div>
				</div>
			</div>
		</div>
	);
};

export default NotesScreen;
