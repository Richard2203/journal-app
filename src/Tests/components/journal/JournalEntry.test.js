// Instalando la dependencia para poder hacer pruebas mock en store de redux
//  https://www.npmjs.com/package/redux-mock-store
// se aÃ±aden las importaciones
//  import configureStore from 'redux-mock-store';
//  const middlewares = [thunk];
//  const mockStore = configureStore(middlewares);
//  const store = mockStore({});

import configureStore from 'redux-mock-store';
//importamos el middleware que usaremos para pruebas
import thunk from 'redux-thunk';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import JournalEntry from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../action/notes';

jest.mock('../../../action/notes', () => ({
	activeNote: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};

// mocksStore recibe objetos para establecer el store; en este caso
// el store contiene el objeto: auth
let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas en el componente <JournalEntry/>', () => {
	const note = {
		id: '12345678',
		date: '12345678',
		title: 'titulo cualquiera',
		body: 'cuerpo cualquiera',
		url: 'https://www.pagina.com',
	};

	const wrapper = mount(
		<Provider store={store}>
			<MemoryRouter>
				<JournalEntry {...note} />
			</MemoryRouter>
		</Provider>
	);

	test('debe renderizar correctamente el componente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe disparar el metodo activeNote', () => {
		wrapper.find('.journal__entry').prop('onClick')();

		// evaluando el llamado del dispatch del store
		expect(store.dispatch).toHaveBeenCalled();
		expect(store.dispatch).toHaveBeenCalledWith(
			activeNote(note.id, { ...note })
		);

		// evaluando el llmado de activeNote
		expect(activeNote).toHaveBeenCalled();
		expect(activeNote).toHaveBeenCalledWith(note.id, { ...note });
	});
});
