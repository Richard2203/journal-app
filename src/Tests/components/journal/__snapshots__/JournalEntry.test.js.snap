// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <JournalEntry/> debe renderizar correctamente el componente 1`] = `
<div
  className="journal__entry cursor animate__animated animate__fadeInDown"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://www.pagina.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      titulo cualquiera
    </p>
    <p
      className="journal__entry-content"
    >
      cuerpo cualquiera
    </p>
  </div>
  <div
    className="journal__entry-date"
  >
    <span>
      Invalid date
    </span>
    <h4>
      Invalid date
    </h4>
  </div>
</div>
`;
